sdks/java/core/src/main/java/org/apache/beam/sdk/io/BigQueryIO.java

```diff
@@ -571,12 +657,535 @@ public boolean getValidate() {
+      executeQuery(
+          queryJobId, query, tableToExtract, flattenResults, bqServices.getJobService(bqOptions));
+      return tableToExtract;
+    }
+
+    private synchronized Job dryRunQueryIfNeeded(BigQueryOptions bqOptions) {
+      if (dryRunJobRef.get() == null) {
+        String projectId = bqOptions.getProject();
+        try {
+          String dryRunJobId = jobIdToken + "-dryRunQuery";
+          Job dryRunJob = dryRunQuery(
+              projectId, dryRunJobId, query, flattenResults, bqServices.getJobService(bqOptions));
+          dryRunJobRef.compareAndSet(null, dryRunJob);
+        } catch (IOException | InterruptedException e) {
+          if (e instanceof InterruptedException) {
+            Thread.currentThread().interrupt();
+          }
+          throw new RuntimeException("Failed to dry run query.", e);
+        }
+      }
+      return dryRunJobRef.get();
+    }
+
+    private static Job dryRunQuery(
+        String projectId,
+        String jobId,
+        String query,
+        boolean flattenResults,
+        JobService jobService) throws InterruptedException, IOException {
+      JobConfigurationQuery queryConfig = new JobConfigurationQuery();
+      queryConfig
+          .setQuery(query)
+          .setCreateDisposition("CREATE_IF_NEEDED")
+          .setFlattenResults(flattenResults)
+          .setPriority("BATCH")
+          .setWriteDisposition("WRITE_EMPTY");
+      jobService.startQueryJob(jobId, queryConfig, true /* dryRun */);
+      Job job = jobService.pollJob(projectId, jobId, QUERY_DRY_RUN_POLL_MAX_RETRIES);
+      if (parseStatus(job) != Status.SUCCEEDED) {
+        throw new IOException("Query job failed: " + jobId);
+      }
+      return job;
+    }
+
+    private static void executeQuery(
+        String jobId,
+        String query,
+        TableReference destinationTable,
+        boolean flattenResults,
+        JobService jobService) throws IOException, InterruptedException {
+      JobConfigurationQuery queryConfig = new JobConfigurationQuery();
+      queryConfig
+          .setQuery(query)
+          .setAllowLargeResults(true)
+          .setCreateDisposition("CREATE_IF_NEEDED")
+          .setDestinationTable(destinationTable)
+          .setFlattenResults(flattenResults)
+          .setPriority("BATCH")
+          .setWriteDisposition("WRITE_EMPTY");
+      jobService.startQueryJob(jobId, queryConfig, false /* dryRun */);
+      Job job = jobService.pollJob(
+          destinationTable.getProjectId(), jobId, JOB_POLL_MAX_RETRIES);
+      if (parseStatus(job) != Status.SUCCEEDED) {
+        throw new IOException("Query job failed: " + jobId);
+      }
+      return;
+    }
+
+    private void readObject(ObjectInputStream in) throws ClassNotFoundException, IOException {
+      in.defaultReadObject();
+      dryRunJobRef = new AtomicReference<>();
+    }
+  }
+
+  abstract static class BigQuerySourceBase extends BoundedSource<TableRow> {
```

Comment:
This PR is very sparse on comments -- this makes it hard to review.

Can you please add some macro-level Javadoc in the important places?

Presumably, here, something like:

```java
/**
 * An abstract {@link BoundedSource} to read a table from BigQuery.
 *
 * <p>This source uses a BigQuery export job to take a snapshot of the table on GCS, and then
 * reads in parallel from each produced file. It is implemented by {@link blahSource}, {@link blah2Source}, and
 * {@link blah3Source}, depending on the configuration of the read and the runner. Specifically,
 *
 * <ul>
 * <li>...which source we use to do what?</li>
 * </ul>
 * ...
 */
```